src/modules/systemlib/bson/tinybson.c

```diff
@@ -338,12 +338,30 @@ write_x(bson_encoder_t encoder, const void *p, size_t s)
 {
 	CODER_CHECK(encoder);
 
-	if (encoder->fd > -1) {
+	/* bson file encoder (non-buffered) */
+	if (encoder->fd > -1 && encoder->buf == NULL) {
 		return (BSON_WRITE(encoder->fd, p, s) == (int)s) ? 0 : -1;
 	}
 
 	/* do we need to extend the buffer? */
 	while ((encoder->bufpos + s) > encoder->bufsize) {
+
+		/* bson buffered file encoder */
+		if (encoder->fd > -1) {
+			// write to disk
+			debug("writing buffer (%d) to disk", encoder->bufpos);
+			int ret = BSON_WRITE(encoder->fd, encoder->buf, encoder->bufpos);
+
+			if (ret == encoder->bufpos) {
+				// reset buffer to beginning and continue
+				encoder->bufpos = 0;
+				break;
```

Comment:
This won't happen in our case, but the buffer could still be too small (`(encoder->bufpos + s) > encoder->bufsize`). breaking here in that case directly leads to a buffer overflow.

Reply:
I added a check that will kill the coder.